File: Algo/Medium/0417_Pacific_Atlantic_Water_Flow/Solution.cpp
// 2022-08-31 (recursive dfs)
class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int n = heights.size();
        int m = heights[0].size();
        
        vector<vector<bool>> canReachPacific(heights.size(), vector<bool>(heights[0].size(), false));
        vector<vector<bool>> canReachAtlantic(heights.size(), vector<bool>(heights[0].size(), false));
        
        // flow upwards from valid cells
        for (int i=0; i<n; i++) {
            flowUpwards(heights, canReachPacific, i, 0);
            flowUpwards(heights, canReachAtlantic, i, m-1);
        }
        for (int i=0; i<m; i++) {
            flowUpwards(heights, canReachPacific, 0, i);
            flowUpwards(heights, canReachAtlantic, n-1, i);
        }
        
        
        vector<vector<int>> output;
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (canReachPacific[i][j] && canReachAtlantic[i][j]) 
                    output.push_back({i, j});
            }
        }
        return output;
    }
    
    void flowUpwards(vector<vector<int>>& heights, vector<vector<bool>>& canReach, int i, int j) {
        if (canReach[i][j])
            return;
            
        vector<vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        canReach[i][j] = true;
        
        for (vector<int> dir : directions) {
            if (i+dir[0] < 0 || i+dir[0] >= heights.size() || j+dir[1] < 0 || j+dir[1] >= heights[i].size()) 
                continue;
            
            if (heights[i][j] <= heights[i+dir[0]][j+dir[1]])
                flowUpwards(heights, canReach, i+dir[0], j+dir[1]);
        }
    } 
};

File: Algo/Medium/0417_Pacific_Atlantic_Water_Flow/Solution.py
from typing import *

## 2021-11-26 (memoized DFS)
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        pacific = list(([False]*len(heights[i])) for i in range(0, len(heights)))
        atlantic = list(([False]*len(heights[i])) for i in range(0, len(heights)))
        
        self.find_flowables(heights, pacific, atlantic)
        return self.find_intersections(pacific, atlantic)
        
    def find_flowables(self, heights, pacific, atlantic):
        for i in range(0, len(heights)):
            for j in range(0, len(heights[i])):
                if i == 0 or j == 0:
                    self.flow_up(heights, atlantic, i, j)
                
                if i == len(heights)-1 or j == len(heights[i])-1:
                    self.flow_up(heights, pacific, i, j)
 
    def flow_up(self, heights, can_reach, i, j):
        current = heights[i][j]
        can_reach[i][j] = True
        
        if i+1 < len(heights) and not can_reach[i+1][j] and current <= heights[i+1][j]:
            heights[i+1][j]
            self.flow_up(heights, can_reach, i+1, j)
        if i-1 >= 0 and not can_reach[i-1][j] and current <= heights[i-1][j]:
            heights[i-1][j]
            self.flow_up(heights, can_reach, i-1, j)
            
        if j+1 < len(heights[0]) and not can_reach[i][j+1] and current <= heights[i][j+1]:
            heights[i][j+1]
            self.flow_up(heights, can_reach, i, j+1)
        if j-1 >= 0 and not can_reach[i][j-1] and current <= heights[i][j-1]:
            heights[i][j-1]
            self.flow_up(heights, can_reach, i, j-1)
   
    def find_intersections(self, pacific, atlantic):
        output = []
        for i in range(0, len(pacific)):
            for j in range(0, len(pacific[i])):
                if atlantic[i][j] and pacific[i][j]:
                    output.append([i, j])
        return output
